simpleclient_httpserver/src/main/java/io/prometheus/client/exporter/HTTPServer.java

```diff
@@ -107,34 +110,109 @@ protected static boolean shouldUseCompression(HttpExchange exchange) {
         server.createContext("/", mHandler);
         server.createContext("/metrics", mHandler);
-        server.setExecutor(Executors.newFixedThreadPool(5));
-        server.start();
+        server.setExecutor(Executors.newFixedThreadPool(5, CustomThreadFactory.defaultThreadFactory(daemon)));
+        start(daemon);
+    }
+
+    /**
+     * Start a HTTP server serving Prometheus metrics from the given registry using non-daemon threads.
+     */
+    public HTTPServer(InetSocketAddress addr, CollectorRegistry registry) throws IOException {
+        this(addr, registry, false);
     }
 
     /**
      * Start a HTTP server serving the default Prometheus registry.
      */
+    public HTTPServer(int port, boolean daemon) throws IOException {
+        this(new InetSocketAddress(port), CollectorRegistry.defaultRegistry, daemon);
+    }
+
+    /**
+     * Start a HTTP server serving the default Prometheus registry using non-daemon threads.
+     */
     public HTTPServer(int port) throws IOException {
-        this(new InetSocketAddress(port), CollectorRegistry.defaultRegistry);
+        this(port, false);
     }
 
     /**
      * Start a HTTP server serving the default Prometheus registry.
      */
+    public HTTPServer(String host, int port, boolean daemon) throws IOException {
+        this(new InetSocketAddress(host, port), CollectorRegistry.defaultRegistry, daemon);
+    }
+
+    /**
+     * Start a HTTP server serving the default Prometheus registry using non-daemon threads.
+     */
     public HTTPServer(String host, int port) throws IOException {
-        this(new InetSocketAddress(host, port), CollectorRegistry.defaultRegistry);
+        this(new InetSocketAddress(host, port), CollectorRegistry.defaultRegistry, false);
+    }
+
+    /**
+     * Start a HTTP server by making sure that its background thread inherit proper daemon flag.
+     */
+    private void start(boolean daemon) {
+        if (daemon == Thread.currentThread().isDaemon()) {
+            server.start();
+        } else {
+            FutureTask<Void> startTask = new FutureTask<Void>(new Runnable() {
+                @Override
+                public void run() {
+                    server.start();
+                }
+            }, null);
+            CustomThreadFactory.defaultThreadFactory(daemon).newThread(startTask).start();
+            try {
+                startTask.get();
+            } catch (InterruptedException e) {
+                //should not happen. but propagate interrupted flag just in case.
+                Thread.currentThread().interrupt();
+            } catch (ExecutionException e) {
+                Throwable cause = e.getCause();
+                if (cause instanceof RuntimeException) {
+                    throw (RuntimeException) cause;
+                } else if (cause instanceof Error) {
+                    throw (Error)cause;
+                } else throw new RuntimeException("Unexpected exception on starting HttpSever", cause);
```

Comment:
The class is HTTPServer

Reply:
This message related to exception on starting com.sun.net.httpserver.HttpServer, but if you insist i will change.